### ISSUE TITLE
Make Unparser more flexible

### ISSUE THREAD

**Issue by vvysotskyi** [janino-compiler/janino#61]
Hello,

Apache Drill widely uses Janino for code compilation, rewriting source code of user-defined functions to be able to use it in generated classes and for some other things.

The main modification which we are doing with user-defined functions is wrapping methods code into the block with the label and replacing all return statements in this method by break ones with mentioned above labels.

We use a class, extended from `Unparser` for this goal, but some of its methods are private, and some of them call visitors.

It would be great if `Unparser.unparseMethodDeclarator()` and `Unparser.unparseStatements()` methods became protected and logic from `Unparser.blockStatementUnparser.visitReturnStatement()` method is moved to protected method in `Unparser` which will be called from `Unparser.blockStatementUnparser.visitReturnStatement()`, so inheritors will be able to extend its logic (similar to the `Unparser.blockStatementUnparser`).

**Comment by aunkrig** [comment]
Hi,

while it would be relatively easy to refactor the `Unparser` as you propose, I have a strong feeling that your approach is wrong: You should not modify the code while you're *unparsing* it, but should instead use a custom `AbstractTraverser` to modify the AST, between parsing and unparsing.

If you cannot get it working, then please provide an example before/after code and I can look into it.


CU Arno

**Comment by aunkrig** [comment]
In other words: The Unparser is not intended to be flexible, actually it should be FINAL.

**Comment by vvysotskyi** [comment]
Thanks for the advice. Just to clarify: saying modification, I meant storing the code that does not correspond to the original statements to the `PrintWriter`.
We collect only the code of method bodies received from unparsing `MethodDeclarator`.
For example, the initial code of the `MethodDeclarator` to be rewritten:
```
public void eval() {
    if (in.isSet == 0 || in.end == in.start) {
        out.isSet = 0;
        return;
    }
    out.isSet = 1;
    out.start = 0;
    out.scale = scale.value;
    out.precision = precision.value;

    byte[] buf = new byte[in.end - in.start];

    in.buffer.getBytes(in.start, buf, 0, in.end - in.start);

    String s = new String(buf, com.google.common.base.Charsets.UTF_8);
    java.math.BigDecimal bd = new java.math.BigDecimal(s);

    org.apache.drill.exec.util.DecimalUtility.checkValueOverflow(bd, precision.value, scale.value);
    bd = bd.setScale(scale.value, java.math.RoundingMode.HALF_UP);

    byte[] bytes = bd.unscaledValue().toByteArray();
    int len = bytes.length;

    out.buffer = buffer.reallocIfNeeded(len);
    out.buffer.setBytes(out.start, bytes);
    out.end = out.start + len;
}
```
Code after the modification:
```
CastEmptyStringNullableVarCharToNullableVarDecimal_eval: {
    if (in.isSet == 0 || in.end == in.start) {
        out.isSet = 0;
        break CastEmptyStringNullableVarCharToNullableVarDecimal_eval;
    }
    out.isSet = 1;
    out.start = 0;
    out.scale = scale.value;
    out.precision = precision.value;

    byte[] buf = new byte[in.end - in.start];

    in.buffer.getBytes(in.start, buf, 0, in.end - in.start);

    String s = new String(buf, com.google.common.base.Charsets.UTF_8);
    java.math.BigDecimal bd = new java.math.BigDecimal(s);

    org.apache.drill.exec.util.DecimalUtility.checkValueOverflow(bd, precision.value, scale.value);
    bd = bd.setScale(scale.value, java.math.RoundingMode.HALF_UP);

    byte[] bytes = bd.unscaledValue().toByteArray();
    int len = bytes.length;

    out.buffer = buffer.reallocIfNeeded(len);
    out.buffer.setBytes(out.start, bytes);
    out.end = out.start + len;
}
```
Could you please share an example, how to modify the tree using `AbstractTraverser`?

I have noticed that all methods in `AbstractTraverser` do not return any value, so, for example, if we want to replace `return` statement by `break`, we can do it only in the method, where the list which contains the concrete statement to replace is used. But in this case, every statement should be checked, whether it should be replaced.

**Comment by aunkrig** [comment]
@vvysotskyi I coincidentially noticed that DRILL implements a `AloadPopRemover` class to work around a Janino-related problem. I added a small optimization in `UnitCompiler` that should render that workaround unnecessary. Could you please test?
See #62.

**Comment by aunkrig** [comment]
I looked into https://github.com/apache/drill/branches/1.14.0/exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/MethodGrabbingVisitor.java and hope that `traverseMethodDeclarator()` is the relevant method...

I see that in your case, a DOM analysis and modification is not necessary.

I recommend that you replace
```java
    ModifiedUnparseVisitor v = new ModifiedUnparseVisitor(writer);
    md.accept(v);
```
with
```java
    UnparseVisitor uv = new UnparseVisitor(writer);
    writer.println(md.name + ": {");
    for (Java.BlockStatement bs : md.optionalStatements) bs.accept(uv);
    writer.println("}");
```

**Comment by vvysotskyi** [comment]
Yes, this is the right place. Currently, Drill 1.14.0 and master version use Janino 2.7.6, but there is a PR https://github.com/apache/drill/pull/1503, where we are moving to the latest version.

As I mentioned above, we also need to replace all `return` statements by `break` ones, therefore we need a DOM analysis and modification.

**Comment by aunkrig** [comment]
Hey Volodymyr,

I added a simple framework that you can use to modify an AST in a very elegant way.

`AstTest.testMethodToLabeledStatement()` implements exactly the test case that you provided.

Please review!

```java

        // Parse the method and get its body.
        MethodDeclarator md1 = new Parser(new Scanner(null, new StringReader(text1))).parseMethodDeclaration();
        List<? extends BlockStatement> ss = md1.optionalStatements;
        assert ss != null;

        // Now generate a "labeled statement".
        LabeledStatement ls2;
        {
            final String label = "CastEmptyStringNullableVarCharToNullableVarDecimal_" + md1.name;
            Block b = new Block(md1.getLocation());
            b.addStatements(new DeepCopier() {

                @Override public BlockStatement
                copyReturnStatement(ReturnStatement subject) {
                    return new Java.BreakStatement(subject.getLocation(), label);
                }

            }.copyBlockStatements(ss));
            ls2 = new LabeledStatement(md1.getLocation(), label, b);
        }

        // Unparse the labeled statement.
        String actual;
        {
            StringWriter sw = new StringWriter();
            Unparser unparser = new Unparser(sw);
            unparser.unparseBlockStatement(ls2);
            unparser.close();
            actual = sw.toString();
        }
```

**Comment by vvysotskyi** [comment]
Hi Arno,

Thanks a lot for adding `DeepCopier` and specifying the example. I have checked and it works!
The single thing is that in `DeepCopier` was added the import to the `AstTest` and it caused the compilation error because test sources are unavailable for the non-test scope.

**Comment by aunkrig** [comment]
Version 3.0.11 is out the door, and includes the fix for this issue.
